New integration request: PZEM-004

<!-- READ THIS FIRST:
 - This is for new integrations (such as new sensors/modules) only, for new features within the environment please use the "feature request" template.
 - Do not delete anything from this template and fill out the form as precisely as possible.
-->

**What new integration would you wish to have?**
- Peacefair PZEM-004
- Peacefair PZEM-004T (same, same but without housing and display)

**If possible, provide a link to an existing library for the integration:**
- https://github.com/olehs/PZEM004T
- https://halfforked.wordpress.com/2018/05/18/low-cost-house-power-monitoring-with-pzem-004-for-home-assistant/
- http://pdacontrolen.com/electricity-consumption-meter-peacefair-pzem-004-esp8266-arduino-nano/
- https://www.instructables.com/id/Power-Peacefair-PZEM-004-ESP8266-Arduino-Nano/

**Product link:**
- https://www.aliexpress.com/item/ANENG-4in1-Digital-Ammeter-Voltmeter-Watt-Power-Energy-Meter-AC-80-260V-100A-PZEM-004/32847092493.html

**Is your feature request related to a problem? Please describe.**
I'm always frustrated when I can't use esphomelib for my integrations :cry: 

**Additional context**
Want to measure the total energy consumption/load of a one-phase main connection in real time with esphomelib/yaml :+1:

I just installed the dev version

Got this error

`INFO Reading configuration /config/esphome/pzem.yaml...
INFO Detected timezone 'CET' with UTC offset 1 and daylight savings time from 29 March 02:00:00 to 25 October 03:00:00
INFO Generating C++ source...
INFO Compiling app...
INFO Running:  platformio run -d /config/esphome/pzem
Processing pzem (board: nodemcuv2; framework: arduino; platform: espressif8266@1.8.0)
--------------------------------------------------------------------------------
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
Dependency Graph
|-- <ESPAsyncTCP-esphome> 1.2.2
|   |-- <ESP8266WiFi> 1.0
|-- <ESP8266WiFi> 1.0
|-- <ESP8266mDNS>
|   |-- <ESP8266WiFi> 1.0
Compiling /data/pzem/.pioenvs/pzem/src/main.cpp.o
Linking /data/pzem/.pioenvs/pzem/firmware.elf
/data/pzem/.pioenvs/pzem/src/pzemac.cpp.o: In function `esphome::pzemac::PZEMAC::update()':
pzemac.cpp:(.text._ZN7esphome6pzemac6PZEMAC6updateEv+0x4): multiple definition of `esphome::pzemac::PZEMAC::update()'
/data/pzem/.pioenvs/pzem/src/esphome/components/pzemac/pzemac.cpp.o:pzemac.cpp:(.text._ZN7esphome6pzemac6PZEMAC6updateEv+0x4): first defined here
collect2: error: ld returned 1 exit status
*** [/data/pzem/.pioenvs/pzem/firmware.elf] Error 1
========================= [FAILED] Took 16.99 seconds =========================`

my yaml file is


`esphome:
  name: pzem
  platform: ESP8266
  board: nodemcuv2
  arduino_version: 2.4.2


wifi:
  ssid: !secret esphome_ssid
  password: !secret esphome_password

# Enable logging
logger:

# Enable Home Assistant API
api:

ota:

uart:
  tx_pin: GPIO1
  rx_pin: GPIO3
  baud_rate: 9600
  id: pzem004t1

time:
  platform: sntp
  id: my_time
  
sensor:
  - platform: pzemac
    current:
      name: "enedis_courant"
    voltage:
      name: "enedis_tension"
    power:
      name: "enedis_puissance"
    frequency:
      name: "enedis_frequence"
    power_factor:
      name: "enedis_facteur_puissance"
    update_interval: 5s

 
switch:
  - platform: restart
    name: "Power Monitor Restart"
  - platform: uart
    uart_id: pzem004t1
    name: "enedis_reset"
    data: [0x01, 0x42, 0x80, 0x11]  `

--- a/esphome/components/modbus/modbus.cpp
+++ b/esphome/components/modbus/modbus.cpp
@@ -43,7 +43,7 @@ uint16_t crc16(uint8_t *data, uint8_t len) {
 bool Modbus::parse_modbus_byte_(uint8_t byte) {
   size_t at = this->rx_buffer_.size();
   this->rx_buffer_.push_back(byte);
-  uint8_t *raw = &this->rx_buffer_[0];
+  const uint8_t *raw = &this->rx_buffer_[0];
 
   // Byte 0: modbus address (match all)
   if (at == 0)
@@ -69,7 +69,7 @@ bool Modbus::parse_modbus_byte_(uint8_t byte) {
     return true;
   // Byte 3+len+1: CRC_HI (over all bytes)
   uint16_t computed_crc = crc16(raw, 3 + data_len);
-  uint16_t remote_crc = uint16_t(raw[3 + data_len]) | (uint16_t(raw[3 + data_len]) << 8);
+  uint16_t remote_crc = uint16_t(raw[3 + data_len]) | (uint16_t(raw[3 + data_len + 1]) << 8);
   if (computed_crc != remote_crc) {
     ESP_LOGW(TAG, "Modbus CRC Check failed! %02X!=%02X", computed_crc, remote_crc);
     return false;

--- a/esphome/components/modbus/modbus.cpp
+++ b/esphome/components/modbus/modbus.cpp
@@ -24,7 +24,7 @@ void Modbus::loop() {
   }
 }
 
-uint16_t crc16(uint8_t *data, uint8_t len) {
+uint16_t crc16(const uint8_t *data, uint8_t len) {
   uint16_t crc = 0xFFFF;
   while (len--) {
     crc ^= *data++;

